sky/adaptors/aws.py

```diff
@@ -98,21 +98,31 @@ def resource(service_name: str, **kwargs):
 
     Args:
         service_name: AWS resource name (e.g., 's3').
-        kwargs: Other options.
+        kwargs: Other options. We add max_attempts to the kwargs instead of
+            using botocore.config.Config() because the latter is not hashable.
     """
+    assert 'config' not in kwargs, (
+        'config should not be passed to resource(), '
+        'as botocore.config.Config() is not hashiable.')
     if not hasattr(_local, 'resource'):
+        # TODO(zhwu): this dict should be LRU cache, instead of growing forever,
+        # to avoid issues in #2668.
         _local.resource = {}
 
     # Using service name and kwargs as key
     sorted_kwargs = tuple(sorted(kwargs.items(), key=lambda x: x[0]))
     key = (service_name, sorted_kwargs)
+    max_attempts = kwargs.pop('max_attempts', None)
+    if max_attempts is not None:
+        config = botocore_config().Config(
+            retries={'max_attempts': max_attempts})
+        kwargs['config'] = config
     if key not in _local.resource:
         with _session_creation_lock:
-            # NOTE: we need the lock here to avoid
-            # thread-safety issues when creating the resource,
-            # because Python module is a shared object,
-            # and we are not sure if the code inside
-            # 'session().resource()' is thread-safe.
+            # NOTE: we need the lock here to avoid thread-safety issues when
+            # creating the resource, because Python module is a shared object,
+            # and we are not sure if the code inside 'session().resource()'
+            # is thread-safe.
             _local.resource[key] = session().resource(service_name, **kwargs)
```

Comment:
we should set a maximum cache size (like LRU cache) to prevent further OOM. but I do not know how hard it is to implement it. maybe we can look into python LRU cache implementation (or can we just use lru_cache after all these changes?)

Reply:
Switched back to `functools.lru_cache` now. Please take another look @suquark